File: front/src/models/resources/movie.ts
import { z } from "zod";
import { ImagesP, ResourceP } from "../traits";
import { zdate } from "../utils";
import { CollectionP } from "./collection";
import { Genre } from "./genre";
import { MetadataP } from "./metadata";
import { Status } from "./show";
import { StudioP } from "./studio";
import { WatchStatusP } from "./watch-status";

export const MovieP = ResourceP("movie")
	.merge(ImagesP)
	.extend({
		/**
		 * The title of this movie.
		 */
		name: z.string(),
		/**
		 * A catchphrase for this show.
		 */
		tagline: z.string().nullable(),
		/**
		 * The list of alternative titles of this movie.
		 */
		aliases: z.array(z.string()),
		/**
		 * The summary of this movie.
		 */
		overview: z.string().nullable(),
		/**
		 * A list of tags that match this movie.
		 */
		tags: z.array(z.string()),
		/**
		 * /** Is this movie not aired yet or finished?
		 */
		status: z.nativeEnum(Status),
		/**
		 * How well this item is rated? (from 0 to 100).
		 */
		rating: z.number().int().gte(0).lte(100),
		/**
		 * How long is this movie? (in minutes).
		 */
		runtime: z.number().int().nullable(),
		/**
		 * The date this movie aired. It can also be null if this is unknown.
		 */
		airDate: zdate().nullable(),
		/**
		 * A youtube url for the trailer.
		 */
		trailer: z.string().optional().nullable(),
		/**
		 * The list of genres (themes) this movie has.
		 */
		genres: z.array(z.nativeEnum(Genre)),
		/**
		 * The studio that made this movie.
		 */
		studio: StudioP.optional().nullable(),
		/**
		 * The collection this movie is part of.
		 */
		collections: z.array(CollectionP).optional(),
		/**
		 * The links to see a movie or an episode.
		 */
		links: z.object({
			/**
			 * The direct link to the unprocessed video (pristine quality).
			 */
			direct: z.string(),

			/**
			 * The link to an HLS master playlist containing all qualities available for this video.
			 */
			hls: z.string().nullable(),
		}),
		/**
		 * The link to metadata providers that this show has.
		 */
		externalId: MetadataP,
		/**
		 * Metadata of what an user as started/planned to watch.
		 */
		watchStatus: WatchStatusP.optional().nullable(),
	})
	.transform((x) => ({
		...x,
		runtime: x.runtime === 0 ? null : x.runtime,
	}))
	.transform((x) => {
		if (!x.thumbnail && x.poster) {
			x.thumbnail = { ...x.poster };
			if (x.thumbnail) {
				x.thumbnail.low = x.thumbnail.high;
				x.thumbnail.medium = x.thumbnail.high;
			}
		}
		return x;
	})
	.transform((x) => ({
		...x,
		href: `/movie/${x.slug}`,
		playHref: `/movie/${x.slug}/watch`,
	}));

/**
 * A Movie type
 */
export type Movie = z.infer<typeof MovieP>;

File: scanner/providers/types/movie.py
from dataclasses import asdict, dataclass, field
from datetime import date
from typing import Optional
from enum import Enum

from providers.utils import select_translation, select_image

from .collection import Collection
from .genre import Genre
from .studio import Studio
from .metadataid import MetadataID


class Status(str, Enum):
	UNKNOWN = "unknown"
	FINISHED = "finished"
	PLANNED = "planned"


@dataclass
class MovieTranslation:
	name: str
	tagline: Optional[str] = None
	tags: list[str] = field(default_factory=list)
	overview: Optional[str] = None

	posters: list[str] = field(default_factory=list)
	logos: list[str] = field(default_factory=list)
	trailers: list[str] = field(default_factory=list)
	thumbnails: list[str] = field(default_factory=list)


@dataclass
class Movie:
	original_language: Optional[str]
	aliases: list[str]
	air_date: Optional[date | int]
	status: Status
	rating: int
	runtime: Optional[int]
	studios: list[Studio]
	genres: list[Genre]
	# TODO: handle staff
	# staff: list[Staff]
	external_id: dict[str, MetadataID]

	path: Optional[str] = None
	# The title of this show according to it's filename (None only for ease of use in providers)
	file_title: Optional[str] = None
	collections: list[Collection] = field(default_factory=list)
	translations: dict[str, MovieTranslation] = field(default_factory=dict)

	def to_kyoo(self):
		trans = select_translation(self) or MovieTranslation(name=self.file_title or "")
		return {
			**asdict(self),
			**asdict(trans),
			"poster": select_image(self, "posters"),
			"thumbnail": select_image(self, "thumbnails"),
			"logo": select_image(self, "logos"),
			"trailer": select_image(self, "trailers"),
			"studio": next((x.to_kyoo() for x in self.studios), None),
			"genres": [x.to_kyoo() for x in self.genres],
			"collections": None,
			"file_title": None,
		}
